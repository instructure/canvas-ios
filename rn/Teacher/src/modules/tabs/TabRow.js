//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/**
* Launching pad for navigation for a single course
* @flow
*/

import React from 'react'
import { SafeAreaView, Image } from 'react-native'
import i18n from 'format-message'

import Images from '../../images'
import Row from '../../common/components/rows/Row'

type Props = {
  tab: Tab,
  color: string,
  onPress: Function,
  attendanceTabID: ?string,
  selected: ?boolean,
}

export default class TabRow extends React.Component<Props> {
  onPress = () => {
    const tab = this.props.tab
    this.props.onPress(tab)
  }

  image = () => {
    const tab = this.props.tab
    let image = Images.course[tab.id]
    if (tab.id === this.props.attendanceTabID) {
      return Images.course.attendance
    }
    if (tab.id.match(/external_tool/)) {
      return Images.course.lti
    }

    if (!image) {
      image = Images.course.placeholder
    }

    return image
  }

  render () {
    const tab = this.props.tab
    return (<SafeAreaView>
      <Row
        title={tab.label}
        image={this.image()}
        imageTint={this.props.color}
        imageSize={{ height: 24, width: 24 }}
        onPress={this.onPress}
        disclosureIndicator
        border='bottom'
        selected={this.props.selected}
        testID={`courses-details.${tab.id}-cell`}
        accessories={tab.hidden &&
            <Image source={Images.invisible} accessibilityLabel={i18n('Hidden')} />
        }
      />
    </SafeAreaView>)
  }
}
