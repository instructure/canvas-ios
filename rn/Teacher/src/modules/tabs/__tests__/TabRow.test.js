//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import { shallow } from 'enzyme'
import React from 'react'
import TabRow from '../TabRow'
import Images from '../../../images'

const template = {
  ...require('../../../__templates__/tab'),
}

const defaultProps = {
  tab: template.tab(),
  courseColor: 'white',
  onPress: () => {},
}

describe('TabRow', () => {
  it('renders correctly', () => {
    const tree = shallow(<TabRow {...defaultProps} />)
    expect(tree).toMatchSnapshot()
  })

  it('shows a default icon', () => {
    const props = {
      ...defaultProps,
      tab: template.tab({ id: 'test default icon' }),
    }
    const tree = shallow(<TabRow {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('can be selected', () => {
    const tree = shallow(<TabRow {...defaultProps} selected />)
    expect(tree).toMatchSnapshot()
  })

  it('can be pressed', () => {
    const onPressed = jest.fn()
    const props = {
      ...defaultProps,
      onPress: onPressed,
    }

    const tree = shallow(<TabRow {...props} />)
    expect(tree).toMatchSnapshot()
    tree.find('Row').simulate('press')
    expect(onPressed).toHaveBeenCalledTimes(1)
  })

  it('uses attendance tab image', () => {
    const props = {
      ...defaultProps,
      attendanceTabID: defaultProps.tab.id,
    }
    const tree = shallow(<TabRow {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('uses attendance lti image', () => {
    const props = {
      ...defaultProps,
      tab: template.tab({ id: '/external_tool/' }),
    }
    const tree = shallow(<TabRow {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders hidden if hidden', () => {
    const props = {
      ...defaultProps,
      tab: template.tab({ id: 'files', hidden: true }),
    }
    const tree = shallow(<TabRow {...props} />)
    const accessories = shallow(tree.find('Row').prop('accessories'))
    expect(accessories.prop('source')).toEqual(Images.invisible)
  })

  it('does not render hidden icon not hidden', () => {
    const props = {
      ...defaultProps,
      tab: template.tab({ id: 'files', hidden: false }),
    }
    const tree = shallow(<TabRow {...props} />)
    expect(tree.find('Row').prop('accessories')).toBeFalsy()
  })
})
