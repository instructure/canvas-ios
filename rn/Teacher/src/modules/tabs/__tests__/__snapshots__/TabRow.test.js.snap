// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabRow can be pressed 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/Assignment.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</SafeAreaView>
`;

exports[`TabRow can be selected 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/Assignment.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    selected={true}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</SafeAreaView>
`;

exports[`TabRow renders correctly 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/Assignment.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</SafeAreaView>
`;

exports[`TabRow shows a default icon 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/Placeholder.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    testID="courses-details.test default icon-cell"
    title="Home"
  />
</SafeAreaView>
`;

exports[`TabRow uses attendance lti image 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/LTI.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    testID="courses-details./external_tool/-cell"
    title="Home"
  />
</SafeAreaView>
`;

exports[`TabRow uses attendance tab image 1`] = `
<SafeAreaView>
  <Row
    border="bottom"
    disclosureIndicator={true}
    image={
      Object {
        "testUri": "../../../src/images/course/attendance-icon.png",
      }
    }
    imageSize={
      Object {
        "height": 24,
        "width": 24,
      }
    }
    onPress={[Function]}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</SafeAreaView>
`;
